src/gui/activityjob.cpp

```diff
@@ -0,0 +1,147 @@
+#include <QUrlQuery>
+#include <QJsonArray>
+
+#include "activityjob.h"
+#include "iconjob.h"
+#include "tray/ActivityData.h"
+
+Q_LOGGING_CATEGORY(lcActivityJob, "nextcloud.gui.activityjob", QtInfoMsg)
+
+namespace OCC {
+
+ActivityJob::ActivityJob(QObject *parent)
+    : QObject(parent)
+{
+}
+
+OcsActivityJob::OcsActivityJob(AccountPtr account, QObject *parent)
+    : ActivityJob(parent)
+    , _account(account)
+{
+}
+
+void OcsActivityJob::queryActivities(const Optional<QString> &objectType, const Optional<QString> &objectId, int limit)
+{
+    _limit = limit;
+    _objectId = objectId;
+    _objectType = objectType;
+
+    startJsonApiJob();
+}
+
+void OcsActivityJob::startJsonApiJob(const Optional<QString> since)
+{
+    const auto url = _objectType || _objectId ? QStringLiteral("ocs/v2.php/apps/activity/api/v2/activity/filter")
+                                              : QStringLiteral("ocs/v2.php/apps/activity/api/v2/activity");
+    auto job = new JsonApiJob(_account, url, this);
+    QObject::connect(job, &JsonApiJob::jsonReceived,
+        this, [this, job](const QJsonDocument &json, int statusCode) {
+            jsonApiJobFinished(*job, json, statusCode);
+        });
+
+    QUrlQuery params;
+    if (_objectType) {
+        params.addQueryItem(QStringLiteral("object_type"), *_objectType);
+    }
+    if (_objectId) {
+        params.addQueryItem(QStringLiteral("object_id"), *_objectId);
+    }
+    if (since) {
+        params.addQueryItem(QStringLiteral("since"), *since);
+    }
+    params.addQueryItem(QStringLiteral("limit"), QString::number(_limit));
+    job->addQueryParams(params);
+    job->start();
+}
+
+Activity OcsActivityJob::jsonObjectToActivity(QJsonObject activityJson)
+{
+    Activity activity;
+    activity._type = Activity::ActivityType;
+    activity._objectType = activityJson.value("object_type").toString();
+    activity._accName = _account->displayName();
+    activity._id = activityJson.value("activity_id").toInt();
+    activity._fileAction = activityJson.value("type").toString();
+    activity._subject = activityJson.value("subject").toString();
+    activity._message = activityJson.value("message").toString();
+    activity._file = activityJson.value("object_name").toString();
+    activity._link = QUrl(activityJson.value("link").toString());
+    activity._dateTime = QDateTime::fromString(activityJson.value("datetime").toString(), Qt::ISODate);
+    activity._icon = activityJson.value("icon").toString();
+
+    return activity;
+}
+
+std::vector<Activity> OcsActivityJob::jsonArrayToActivities(QJsonArray activitiesJson)
+{
+    std::vector<Activity> activities;
+    for (const auto activ : activitiesJson) {
+        activities.emplace_back(jsonObjectToActivity(activ.toObject()));
```

Comment:
Yes, I know, but I thought let's use `emplace_back` instead of `push_back`, because it can do everything that `push_back` can + some extras. Your comment made me realise that there might occur bugs. https://stackoverflow.com/a/36919571/6111101